zqd/api/connection.go

```diff
@@ -0,0 +1,261 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path"
+	"strconv"
+
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio/bzngio"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/go-resty/resty/v2"
+)
+
+const (
+	// DefaultPort zqd port to connect with.
+	DefaultPort = 9867
+)
+
+var (
+	// ErrSpaceNotFound returns when specified space does not exist.
+	ErrSpaceNotFound = errors.New("space not found")
+	// ErrSpaceExists returns when specified the space already exists.
+	ErrSpaceExists = errors.New("space exists")
+)
+
+type Connection struct {
+	client *resty.Client
+}
+
+func newConnection(useragent string, client *resty.Client) *Connection {
+	if useragent != "" {
+		client.SetHeader("User-Agent", useragent)
+	}
+	client.SetError(Error{})
+	client.OnAfterResponse(checkError)
+	return &Connection{
+		client: client,
+	}
+}
+
+// NewConnection creates a new connection with the given useragent string
+// and a base URL set up to talk to http://localhost:defaultport
+func NewConnection(useragent string) *Connection {
```

Comment:
Since setting a non-default user agent will be rare, how about removing it from these constructors and provide a setter?

I'd also suggest that the constructors just take a url string, and not constructors that differ only on url-string vs *url.URL; http.NewRequest takes a `url string` parameter.

Reply:
Good idea.